### PR TITLE
Keep skiplist iterator ordered

### LINKED ISSUE #737: SkipList iterator resumes from the beginning after it reaches the end
Currently, the iterator returned by `SkipMap::iter` resumes from the beginning after it reaches the end. This happens even when there is no concurrent writer:

```rust
fn foo() {
    let map: SkipMap<i32, i32> = iter::once((1, 1)).collect();
    let mut iter = map.iter();
    assert!(iter.next().is_some());
    assert!(iter.next().is_none());
    assert!(iter.next().is_none()); // error
}
```

IMHO, the same iterator instance should not return the same value multiple times as it will break the ordered property of skiplist.

### PR DESCRIPTION
**Pull request by topecongiro**
This PR updates the implementation of `RefIter` so that `head` and `tail` always hold the last value returned from `next` and `next_back` respectively. `None` means there has been no invocation.

This PR also tweaks how we handle reference count in RefIter, and adds `Drop` implementation to `Iter` which is a consumer of `RefIter` to avoid leaking memory.

Closes #737.

### CODE REVIEW

**Review by taiki-e — COMMENTED** [review]
LGTM aside from a nit. Thanks for reporting and fixing this!

File: crossbeam-skiplist/src/base.rs [hunk 1]
```diff
@@ -1708,53 +1708,64 @@ where
...
-            self.head = None;
-            self.tail = None;
-        }
-        self.tail.clone()
+    }
+}
+
+impl<'a, K: 'a, V: 'a> RefIter<'a, K, V> {
+    /// Decrements the reference count of `RefEntry` owned by the iterator.
+    pub fn drop_impl(&mut self, guard: &Guard) {
```
[the remark this answers is not in the document]
Comment (topecongiro) [reply crossbeam-skiplist/src/base.rs:1765]: Thank you for the review. Sounds like a legit concern to me.

With 37e7236bdad72ad4292d0d6a4cbae42bf06891ee it should be safe to call this function multiple times.

[the remark this answers is not in the document]
Comment (taiki-e) [reply crossbeam-skiplist/src/base.rs:1765]: > With [37e7236](https://github.com/crossbeam-rs/crossbeam/commit/37e7236bdad72ad4292d0d6a4cbae42bf06891ee) it should be safe to call this function multiple times.

Thanks. That seems a better solution.

**Review by taiki-e — COMMENTED** [review]
r=me after clippy fixed

bors d+

**Pull request merged**